### PR TITLE
Added commandlets for docker and kubernetes setup

### PR DESCRIPTION
**Pull request by isandesh1986**
implements issue #584

### CODE REVIEW

**Review by hohwille — COMMENTED** [review]
Thanks @isandesh1986 - nice progress. Will we in the end add a commandlet that we can do something like:
```
devon docker setup
devon kubernetes setup
```

Those commandlets could detect OS and then do the according steps required or delegate to *.bat file one windows?

File: scripts/src/main/resources/scripts/setup-kubernetes
Comment (maybeec) [inline scripts/src/main/resources/scripts/setup-kubernetes:13]: please also echo the kubernetes version information after fi

File: scripts/src/main/resources/scripts/setup-docker
Comment (maybeec) [inline scripts/src/main/resources/scripts/setup-docker:14]: please do docker --verion after fi so it will in any case be outputted

**Review by maybeec — CHANGES_REQUESTED** [review]

File: scripts/src/main/resources/scripts/setup-docker
Comment (isandesh1986) [reply scripts/src/main/resources/scripts/setup-docker]: Done

File: scripts/src/main/resources/scripts/setup-kubernetes
Comment (isandesh1986) [reply scripts/src/main/resources/scripts/setup-kubernetes:17]: Done

**Comment by isandesh1986** [comment]
> Thanks @isandesh1986 - nice progress. Will we in the end add a commandlet that we can do something like:
> 
> ```
> devon docker setup
> devon kubernetes setup
> ```
> 
> Those commandlets could detect OS and then do the according steps required or delegate to *.bat file one windows?

Yes Jorg. Taken care now.

File: scripts/src/main/resources/scripts/command/docker
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/docker:2]: even though we do not support other OS than windows yet, could you please alread add OS checks as suggested in last review?
```
if doIsWindows
then
   # ... add your code here
else
  doFail "Sorry, docker-installation support is not yet implemented for your OS. Please install manually or help devonfw-ide to support it for your OS by contributing a pull-request."
fi
```

Further, I suggested to design these commandlets following the philosophy of other commandlets.
So what you are doing here should only happen for `devon docker setup`.
For any other command (excluding `help`) you should delegate to an existing docker installation and pass the arguments to the docker command.
Finally, we always do a setup in a lazy way. So currently I can run
```
devon mvn clean install
```
And if `mvn` is not yet installed, it will be installed first.
At the end (both if mvn was already present or just has been installed) `mvn clean install` is called.

You should be able to check if `docker` command is present in an OS independent way.
Then it can already be used on all OS but only the automatic installation does not work on OS other than windows.

Please have a look at existing commandlets like `mvn`, `npm`, `java`, etc. and follow our existing philosophy.

File: scripts/src/main/resources/scripts/command/kubernetes
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/kubernetes:2]: same as with docker commandlet.

File: scripts/src/main/resources/scripts/setup-docker
Comment (hohwille) [inline scripts/src/main/resources/scripts/setup-docker:5]: will this sudo require a password to be entered by the user? Do we need to prompt something before to explain what kind of password is actually required here to avoid confusion?

**Review by hohwille — CHANGES_REQUESTED** [review]
@isandesh1986 thanks for this great PR 👍 The installation on windows is looking good to me.
However, we need some improvements before this can be merged as if used on Linux or Mac this would give the user a very strange technical error.
Can you please look at my review comments and rework accordingly.
Thanks in advance!

File: scripts/src/main/resources/scripts/command/docker
Comment (maybeec) [reply scripts/src/main/resources/scripts/command/docker:2]: I think as docker or kubernetes are global services, it does not really make sense to proxy the commands by devon command.
I even did not understood why we are providing this anyhow, I am not using it at all and it makes me feel even more to be working with a framework if I am told to use `devon mvn *` instead of `mvn *`.

**Comment by hohwille** [comment]
BTW: when we add new commandlets we also add documentation as asciidoc, and also integrate that into the existing documentaiton.
* add `docker.asciidoc` and `kubernetes.asciidoc` like this example: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/mvn.asciidoc
* add `include` here: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/devonfw-ide-usage.asciidoc
* add `link` here: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/cli.asciidoc

File: scripts/src/main/resources/scripts/setup-docker
Comment (isandesh1986) [reply scripts/src/main/resources/scripts/setup-docker:5]: Added a hint.

File: scripts/src/main/resources/scripts/command/docker
Comment (isandesh1986) [reply scripts/src/main/resources/scripts/command/docker]: > even though we do not support other OS than windows yet, could you please alread add OS checks as suggested in last review?
> 
> ```
> if doIsWindows
> then
>    # ... add your code here
> else
>   doFail "Sorry, docker-installation support is not yet implemented for your OS. Please install manually or help devonfw-ide to support it for your OS by contributing a pull-request."
> fi
> ```
> 
> Further, I suggested to design these commandlets following the philosophy of other commandlets.
> So what you are doing here should only happen for `devon docker setup`.
> For any other command (excluding `help`) you should delegate to an existing docker installation and pass the arguments to the docker command.
> Finally, we always do a setup in a lazy way. So currently I can run
> 
> ```
> devon mvn clean install
> ```
> 
> And if `mvn` is not yet installed, it will be installed first.
> At the end (both if mvn was already present or just has been installed) `mvn clean install` is called.
> 
> You should be able to check if `docker` command is present in an OS independent way.
> Then it can already be used on all OS but only the automatic installation does not work on OS other than windows.
> 
> Please have a look at existing commandlets like `mvn`, `npm`, `java`, etc. and follow our existing philosophy.

Have incorporated the OS checks now.

**Comment by isandesh1986** [comment]
> BTW: when we add new commandlets we also add documentation as asciidoc, and also integrate that into the existing documentaiton.
> 
> * add `docker.asciidoc` and `kubernetes.asciidoc` like this example: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/mvn.asciidoc
> * add `include` here: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/devonfw-ide-usage.asciidoc
> * add `link` here: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/cli.asciidoc

sure, will add.

**Comment by isandesh1986** [comment]
> BTW: when we add new commandlets we also add documentation as asciidoc, and also integrate that into the existing documentaiton.
> 
> * add `docker.asciidoc` and `kubernetes.asciidoc` like this example: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/mvn.asciidoc
> * add `include` here: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/devonfw-ide-usage.asciidoc
> * add `link` here: https://github.com/devonfw/ide/blob/0a7d8a6af55829efb6a2a99fa37ca964aef07c48/documentation/cli.asciidoc

Documentation has been pushed for these commandlets.

File: scripts/src/main/resources/scripts/command/docker
Comment (hohwille) [reply scripts/src/main/resources/scripts/command/docker]: Nobody is forced to do `devon mvn *` instead of `mvn *`. However, this allows scripts to be portable and work seamlessly. Otherwise, you would have to write something like:
```
if ! command -v mvn &> /dev/null
then
  devon mvn setup
fi
mvn my-goal
```

This might still not be enough if `mvn` is installed but the devon environment variables have not been set accordingly.
Instead a script can also do
```
devon mvn my-goal
```
and it will always work and even install maven on the fly if not yet available.

Aint the entire goal of our docker and kubernetes support that projects could also provide scripts to provide e.g. a database running via container as a most simple script in a portal and automated way?

If you want we can have a fundamental discussion about the philopsophy of devonfw-ide and if you think that I designed all odd and wrong. However, we IMHO here want to discuss about adding these two commandlets.
I am repeating myself, but I do not accept that all existing commandlets follow a common philosophy for IMHO good reasons and that these two new commandelts should ignore them what would also is required for specific features to work. E.g. you currently can not add `docker` or `kubernetes` to `DEVON_IDE_TOOLS` variable because of that.

**Review by maybeec — APPROVED** [review]

**Comment by hohwille** [comment]
just in case something is unclear:
This PR needs rework such that the `docker` and `kubernetes` commandlets follow the style of all the other existing commandlets:
* parse arguments
* arg `setup` triggers the current functionality
* arg `help` prints the help
* in any other case call "silent setup" (will do nothing if the executable native command `docker` or `kubernetes` is available (`command -v`). Then delegate args to that command.

Thanks

**Comment by hohwille** [comment]
FYI: I created an official issue for this story:  #584
There I also wrote some hints for MacOS support. I could also provide bash snipplets or if this PR is merged, I can create another PR that would add some basic MacOS support (semi-automated).

**Comment by maybeec** [comment]
@isandesh1986, @ssarmokadam would you be able to provide a fix for that at a short term or do we need to estimate a task for the next sprint to keep track? Please organize on your own.

**Comment by isandesh1986** [comment]
> @isandesh1986, @ssarmokadam would you be able to provide a fix for that at a short term or do we need to estimate a task for the next sprint to keep track? Please organize on your own.

Hi @maybeec , We have created 2 separate tasks for Docker and Kubernetes and placed it in backlog. Based on the capacity available we will pull it into next sprint mostly. Thanks

**Comment by isandesh1986** [comment]
> just in case something is unclear:
> This PR needs rework such that the `docker` and `kubernetes` commandlets follow the style of all the other existing commandlets:
> 
> * parse arguments
> * arg `setup` triggers the current functionality
> * arg `help` prints the help
> * in any other case call "silent setup" (will do nothing if the executable native command `docker` or `kubernetes` is available (`command -v`) and if not present trigger the setup - that will currently fail if not on windows). Then delegate args to that command.
> 
> Thanks

Hi @hohwille , pushed the changes for the commandlet philosophy for Docker-. please review the same. Thanks

File: scripts/src/main/resources/scripts/command/docker
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/docker:19]: Here we should first check if docker command is already available in OS independent way (see other commandlets). If already installed, we are already done.

File: scripts/src/main/resources/scripts/command/kubernetes
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/kubernetes:19]: same here, if `kubectrl` is available as command we can assume that nothing is left to do.

File: scripts/src/main/resources/scripts/command/kubernetes
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/kubernetes:52]: first we need to call `doSetup` to ensure that it is already installed.

File: scripts/src/main/resources/scripts/command/docker
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/docker:52]: here we first should call `doSetup` so that if docker is not already installed, it should first be installed.

File: documentation/kubernetes.asciidoc
Comment (hohwille) [inline documentation/kubernetes.asciidoc:29]: is the last part left from some C&P?
```
" (`«args»`)
```
Looks missplaced to me.

File: scripts/src/main/resources/scripts/command/kubernetes
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/kubernetes:53]: I could never test on windows, but can this run after installation directly or does it have to run via `wsl bash`? If that runs directly this would be perfect and work same for every OS.

File: scripts/src/main/resources/scripts/command/docker
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/docker:53]: can this run directly (without `wsl bash`) on windows?

**Review by hohwille — COMMENTED** [review]
@isandesh1986 thanks for your improvements. This is now perfectly following the philosophy of devonfw-ide and all the other existing commandlets.
Still I have left some last review comments. Also in some places indent is by 4 spaces and could be aligned. Will you be available to update PR once more? Otherwise let me know when I need to adapt the PR directly...

File: scripts/src/main/resources/scripts/command/kubernetes
Comment (hohwille) [reply scripts/src/main/resources/scripts/command/kubernetes:52]: https://github.com/devonfw/ide/blob/master/scripts/src/main/resources/scripts/command/mvn#L309

File: scripts/src/main/resources/scripts/command/docker
Comment (hohwille) [reply scripts/src/main/resources/scripts/command/docker:19]: Example:
https://github.com/devonfw/ide/blob/master/scripts/src/main/resources/scripts/command/ionic#L19

File: scripts/src/main/resources/scripts/command/docker
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/docker:29]: 1. should we also echo a link to the devonfw-ide documentation with further instructions how to install WSL2 that we can even improve after the release?
2. we could move such "WSL" code to `functions` if we want to share it between `docker` and `kubernetes` commandlets.

File: scripts/src/main/resources/scripts/command/docker
Comment (isandesh1986) [reply scripts/src/main/resources/scripts/command/docker:52]: > @isandesh1986 thanks for your improvements. This is now perfectly following the philosophy of devonfw-ide and all the other existing commandlets.
> Still I have left some last review comments. Also in some places indent is by 4 spaces and could be aligned. Will you be available to update PR once more? Otherwise let me know when I need to adapt the PR directly...

Hi @hohwille  - I will update the PR and revert. Thanks

File: scripts/src/main/resources/scripts/command/docker
Comment (isandesh1986) [reply scripts/src/main/resources/scripts/command/docker]: Fixed it by prefixing wsl.

File: scripts/src/main/resources/scripts/command/kubernetes
Comment (isandesh1986) [reply scripts/src/main/resources/scripts/command/kubernetes]: Fixed it by prefixing wsl

File: documentation/kubernetes.asciidoc
Comment (isandesh1986) [reply documentation/kubernetes.asciidoc]: Done. fixed it.

File: scripts/src/main/resources/scripts/command/kubernetes
Comment (isandesh1986) [reply scripts/src/main/resources/scripts/command/kubernetes]: Hi @hohwille, For mvn it does make sense to install mvn when the user calls `mvn install`, but for docker and kubernetes I believe we should either redirect them to call `devon docker/kubernetes setup` rather than directly installing. WDYT?

Option1:- 
`Docker is not installed neither on Windows nor on WSL. Please call devon docker setup to install on WSL.`

Option2:-
`Docker is not installed neither on Windows nor on WSL.  Proceeding to install. Do you want to continue`

File: scripts/src/main/resources/scripts/command/kubernetes
Comment (hohwille) [inline scripts/src/main/resources/scripts/command/kubernetes:55]: why should we limit this to windows?
If `kubectl` is available on linux or MacOS we can make `devon kubernetes «args»` delegate to `kubectl «args»` without errors.
Why not simply reusing `doSetup` as we do in other commandlets? (we specify an argument to `doSetup` so we can distinguish if the user wants to setup explicitly or implicitly).

**Review by hohwille — APPROVED** [review]
@isandesh1986 thanks for your rework and completing this PR.
Now everything is perfect and all review comments are reworked.
I will merge and then add MacOS support.

**Pull request merged**
